torch/lib/c10d/ProcessGroup.hpp

```diff
@@ -119,33 +119,33 @@ class ProcessGroup {
     return size_;
   }
 
-  virtual std::shared_ptr<ProcessGroup::Work> broadcast(
+  virtual c10::intrusive_ptr<ProcessGroup::Work> broadcast(
```

Comment:
This change is probably going to break other third party backends like: https://github.com/intel/torch-ccl/blob/master/src/ProcessGroupCCL.hpp#L136 and https://github.com/openucx/torch-ucc/blob/master/include/torch_ucc.hpp#L77.

I'm guessing this is necessary for TorchScript and there is no way around it, so should we ask the third-party libraries to make this change as well? (we can probably file issues on those repos).

cc @agolynski Since this affects the c10d extension.

Reply:
Yeah agreed that we should ask them to make the changes. Do you have a list of third party backends or are these two the only two that's currently using c10d extension?

Reply:
The ones I am aware of are Intel and UCX. See:

1. https://github.com/openucx/torch-ucc
2. https://github.com/pytorch/pytorch/pull/28068

Could you please check with @agolynski, he might know more context.

Reply:
Hello @chengjunlu @mshiryaev @Sergei-Lebedev @srinivas212 

This PR will break master -> master dependency in torch-ccl and torch-ucc.

In near future we'll be changing ProcessGroup API which will break these repos as well. Would it be okay if you depend on 1.6 (upgrade to 1.7 when released) and not on master meanwhile?

@Sergei-Lebedev @srinivas212: How does torch-ucc depend on pytorch, do you require users to install from pytorch from master branch?

Reply:
Hi @agolynski @mrshenli @pritamdamania87 - after discussions with both Torch-UCC and Torch-CCL teams, the near term plan is to fix the issue in the third-party repo once this change lands.

In general, it is best we keep third-party plugins in sync w/ master.

Reply:
going to land the stack soon, created https://github.com/openucx/torch-ucc/pull/23 and https://github.com/intel/torch-ccl/pull/11 to ucc and ccl to do the API migration.